test/cmocka/src/lib/lib/memset.c

```diff
@@ -0,0 +1,445 @@
+ */
+#include <sof/alloc.h>
+
+#include <stdarg.h>
+#include <setjmp.h>
+#include <cmocka.h>
+
+struct test_bases_ran_status_t {
+	short change_none;
+	short change_first;
+	short change_all;
+	short change_middle;
+};
+
+struct test_data_t {
+	size_t len;
+	char *before;
+	char *after;
+	char *memset_ret;
+};
+
+struct test_data_change_middle_t {
+	size_t beg;
+	size_t end;
+	size_t len;
+};
+
+static struct test_bases_ran_status_t   *test_bases_ran_status;
+static struct test_data_t               *test_data;
+static struct test_data_change_middle_t *test_data_change_middle;
+static const char default_char = 'a';
+
+static void reset_test_arr(void)
+{
+	int i;
+	char *p_bef = test_data->before;
+	char *p_aft = test_data->after;
+
+	for (i = 0; i < test_data->len; ++i)
+		*(p_aft++) = *(p_bef++);
+
+	test_data->memset_ret = NULL;
+}
+
+static int setup(void **state)
+{
+	test_bases_ran_status = malloc(
+		sizeof(struct test_bases_ran_status_t)
+	);
+	test_bases_ran_status->change_none   = 0;
+	test_bases_ran_status->change_first  = 0;
+	test_bases_ran_status->change_all    = 0;
+	test_bases_ran_status->change_middle = 0;
+
+	const int len = 6;
+
+	test_data = malloc(
+		sizeof(struct test_data_t)
+	);
+	test_data->before = malloc(len * sizeof(char));
+	test_data->after  = malloc(len * sizeof(char));
+	test_data->len = len;
+
+	char *p = test_data->before;
+	int i;
+
+	for (i = 0; i < len-1; ++i)
+		*(p++) = default_char;
+
+	*p = 0;
+
+	const int change_beg = 2;
+	const int change_end = 4;
+	// aa000a... - pattern of change
+
+	test_data_change_middle = malloc(
+		sizeof(struct test_data_change_middle_t)
+	);
+	test_data_change_middle->beg = change_beg;
+	test_data_change_middle->end = change_end;
+	test_data_change_middle->len = change_end - change_beg + 1;
+
+	return 0;
+}
+
+static int teardown(void **state)
+{
+	free(test_data->before);
+	free(test_data->after);
+	free(test_data);
+	free(test_data_change_middle);
+	free(test_bases_ran_status);
+
+	return 0;
+}
+
+static int check_arr(const char *arr, const size_t n, const char should_be)
+{
+	size_t i;
+	const char *it = arr;
+
+	for (i = 0; i < n; ++i)
+		if (*(it++) != should_be)
+			return i;
+	return -1;
+}
+static int check_test_arr(const char should_be)
+{
+	return check_arr(
+		test_data->after,
+		test_data->len-1,
+		should_be
+	);
+}
+
+static int check_arrs(const char *arr1, const char *arr2, const size_t n)
+{
+	size_t i;
+	const char *it1 = arr1;
+	const char *it2 = arr2;
+
+	for (i = 0; i < n; ++i)
+		if (*(it1++) != *(it2++))
+			return i;
+	return -1;
+}
+static int check_test_arrs(void)
+{
+	return check_arrs(
+		test_data->before,
+		test_data->after,
+		test_data->len
+	);
+}
+static int check_test_arrs_with_offset(const int offset)
+{
+	return check_arrs(
+		test_data->before + offset,
+		test_data->after  + offset,
+		test_data->len    - offset
+	);
+}
+
+#define test_multipart_base(mem_s, mem_c, mem_n, ran_status) {\
```

Comment:
DRY is good, it would be even better if you could use static functions instead of defines - to have types and not to have spam ugly '\' at newlines, if possible ofc